src/com/amazon/ionelement/api/Element.kt

```diff
@@ -0,0 +1,257 @@
+     * When adding multiple metas, consider [withMetas] instead.
+     */
+    fun withMeta(key: String, value: Any): IonElement = withMetas(metaContainerOf(key to value))
+
+    /** Returns a copy of the current node without any metadata.  (Not recursive.) */
+    fun withoutMetas(): IonElement
+
+    /** Writes the current Ion element to the specified [IonWriter]. */
+    fun writeTo(writer: IonWriter)
+
+    /** Converts the current element to Ion text. */
+    override fun toString(): String
+}
+
+
+/**
+ * Represents an Ion null value.  (Typed or untyped.)
+ *
+ * Includes no additional functionality over [Element], but serves to provide additional type safety when
+ * working with elements that represnt Ion null values..
+ */
+interface NullElement : Element
+
+// TODO: "Bool" or "Boolean"
+/** Represents a Ion bool. */
+interface BooleanElement : Element {
+    val booleanValue: Boolean
+}
+
+/** Represents a Ion timestamp. */
+interface TimestampElement : Element {
+    val timestampValue: Timestamp
+}
+
+/** Represents a Ion int. */
+interface IntElement : Element {
+    val integerSize: IntegerSize
+    val longValue: Long
+    val bigIntegerValue: BigInteger
+}
+
+/** Represents a Ion decimal. */
+interface DecimalElement : Element {
+    val decimalValue: Decimal
+}
+
+/**
+ * Represents a Ion float.
+ *
+ * QUESTION/TODO: "Double" to be consistent with Kotlin or "Float" to be consistent with Ion?
+ */
+interface FloatElement : Element {
+    val doubleValue: Double
+}
+
+
+/** Represents an Ion string or symbol. */
+interface TextElement : Element {
+    val textValue: String
+}
+/**
+ * Represents an Ion string.
+ *
+ * Includes no additional functionality over [TextElement], but serves to provide additional type safety when
+ * working with elements that must be Ion strings.
+ */
+interface StringElement : TextElement
+
+/**
+ * Represents an Ion symbol.
+ *
+ * Includes no additional functionality over [TextElement], but serves to provide additional type safety when
+ * working with elements that must be Ion strings.
+ */
+interface SymbolElement : TextElement
+
+/** Represents an Ion clob or blob. */
+interface LobElement : Element {
+    val bytesValue:  IonByteArray
+}
+/**
+ * Represents an Ion blob.
+ *
+ * Includes no additional functionality over [LobElement], but serves to provide additional type safety when
+ * working with elements that must be Ion blobs.
+ */
+interface BlobElement : LobElement
+
+/**
+ * Represents an Ion clob.
+ *
+ * Includes no additional functionality over [LobElement], but serves to provide additional type safety when
+ * working with elements that must be Ion clobs.
+ */
+interface ClobElement : LobElement
+
+/**
+ * Represents an Ion list, s-expression or struct.
+ *
+ * Items within [values] may or may not be in a defined order.  The order is defined for lists and s-expressions,
+ * but undefined for structs.
+ **/
+interface ContainerElement : Element {
+    val values: Collection<IonElement>
```

Comment:
This was a matter of debate with another member of the team a while back who's argument was that we should use `Collection<T>` where order is undefined and `List<T>` where it is.  This is consistent with the javadoc.  And we're not implementing them directly, just using the default Kotlin implementations, though agreed it would be a pain if we ever did need to use our own implementation.

Also:  I'm also not generally a fan of keeping a `size` property that's not on the `values` collection itself...

```
val containerElem: ContainerElement = /*... */
// seems wrong to me because containerElement doesn't implement Iterable<T>
containerElem.size 

// that's the correct place, I think...
containerElem.values.size 
```

To both reduce the size of the API *and* put `size` at the right place, perhaps we can add another interface:

```
interface SizedIterable<T> : Iterable<T> {
    val size: Int
} 
```

??